### PR TITLE
Systemd shutdown ramfs fixup

### PR DESCRIPTION
**Pull request by ElvishJerricco**
###### Description of changes

The shutdown ramfs wasn't actually doing anything. This fixes it and adds a convenient API for configuring it, along with an example that syncs ZFS pools on shutdown.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ElvishJerricco** [comment]
Should I have it print something like `Syncing ZFS pools...` in case it causes people to worry about their shutdown taking some time?

Also, I *believe* this means we can get rid of the horrible `zfs set` hack here: https://github.com/NixOS/nixpkgs/blob/41b500b737e0ee3d9bfb4e0175d7d3bc88c59111/nixos/modules/tasks/filesystems/zfs.nix#L593-L609

File: nixos/lib/systemd-types.nix [hunk 1]
```diff
@@ -34,4 +34,36 @@ rec {
 
   automounts = with types; listOf (submodule [ stage2AutomountOptions unitConfig automountConfig ]);
   initrdAutomounts = with types; attrsOf (submodule [ stage1AutomountOptions unitConfig automountConfig ]);
+
+  initrdContents = types.attrsOf (types.submodule ({ config, options, name, ... }: {
+    options = {
+      enable = mkEnableOption "copying of this file to initrd and symlinking it" // { default = true; };
```
Comment (dasJ) [inline nixos/lib/systemd-types.nix:40]:
```suggestion
      enable = mkEnableOption "copying of this file and symlinking it" // { default = true; };
```

File: nixos/modules/system/boot/systemd/shutdown.nix [hunk 2]
```diff
@@ -1,31 +1,57 @@
...
 
 in {
-  options.boot.systemd.shutdown = {
+  options.systemd.shutdownRamfs = {
     enable = lib.mkEnableOption "pivoting back to an initramfs for shutdown" // { default = true; };
+    contents = lib.mkOption {
+      description = "Set of files that have to be linked into the shutdown ramfs";
+      example = lib.literalExpression ''
+        {
+          "/lib/systemd/system-shutdown/zpool-export".source = writeShellScript "zpool" "exec ''${zfs}/bin/zpool export -a"
```
Comment (dasJ) [inline nixos/modules/system/boot/systemd/shutdown.nix:17]:
```suggestion
          "/lib/systemd/system-shutdown/zpool-export".source = writeShellScript "zpool" {} "exec ''${zfs}/bin/zpool export -a"
```

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 3]
```diff
@@ -466,6 +466,11 @@ in
         '') rootPools));
       };
 
+      systemd.shutdownRamfs.contents."/etc/systemd/system-shutdown/zpool".source = pkgs.writeShellScript "zpool" ''
```
Comment (dasJ) [inline nixos/modules/tasks/filesystems/zfs.nix:469]:
```suggestion
      systemd.shutdownRamfs.contents."/etc/systemd/system-shutdown/zpool".source = pkgs.writeShellScript "zpool-export-shutdown" ''
```

**Review by dasJ — CHANGES_REQUESTED** [review]
LGTM, only minor nits

File: nixos/modules/system/boot/systemd/shutdown.nix (hunk 2 above)
Comment (ElvishJerricco) [reply nixos/modules/system/boot/systemd/shutdown.nix:17]: Hm? `writeShellScript` only takes two arguments.

Comment (dasJ) [reply nixos/modules/system/boot/systemd/shutdown.nix]: Sorry, I mistook it with `runCommand` :(

**Review by dasJ — APPROVED** [review]

**Comment by flokli** [comment]
So https://github.com/NixOS/nixpkgs/pull/168774#issuecomment-1100748700 was actually not sufficient, and this is the followup fix?

**Comment by dasJ** [comment]
Yes, I'm sorry :/

**Review by flokli — CHANGES_REQUESTED** [review]
Can we update the VM test (and assert for some more outputs that should be visible in stdout)?

**Comment by ElvishJerricco** [comment]
@flokli The test is sort of fixed. It passes if all is well. But since the success condition is `wait_for_console_text` returning, it simply hangs if it should fail. Is that ok to merge, or is there something I can do about it?

**Comment by flokli** [comment]
I think it's fine, as the VM test eventually fails due to timeout.

**Comment by Artturin** [comment]
> I think it's fine, as the VM test eventually fails due to timeout.

then perhaps set `meta.timeout = X;` so that it doesn't take a couple of hours

**Pull request merged**